Deploying CommonConceptsTest causes Invalid character error

When including CommonConceptsTest package in the Rhetos server, the following error is reported from DeployPackages.exe:

```
[Trace] DeployPackages: > adminGui@1.0.0 tsc C:\My Projects\Rhetos\Source\Rhetos\Resources\AdminGuiCompile
[Trace] DeployPackages: > tsc
[Trace] DeployPackages:
[Trace] DeployPackages: scripts/models/rhetos.angular2.ts(21156,97): error TS1127: Invalid character.
```

The generated code in *rhetos.angular2.ts* is:

```typescript
21144    getValidators(): { [propName: string]: ValidatorDefinition[]; } {
21145        return {
21146            'StringFrom200To249':[
21147                {Validator: RegexValidatorFactory('2[0-4][0-9]'), ErrorCode: 'notMatchingRegex', ErrorMessage: 'StringFrom200To249 must be between 200 and 249.' }
21148            ],
21149            'UnicodeTest':[
21150                {Validator: RegexValidatorFactory('čćČĆテスト'), ErrorCode: 'notMatchingRegex', ErrorMessage: 'UnicodeTest must match unicode regex.' }
21151            ],
21152            'WhitespaceTest':[
21153                {Validator: RegexValidatorFactory('a\r\nb\tc'), ErrorCode: 'notMatchingRegex', ErrorMessage: 'WhitespaceTest must match whitespace regex.' }
21154            ],
21155            'SpecialCharTest':[
21156                {Validator: RegexValidatorFactory('a\!\@\#\$\%\^\&\*\(\)_\+\-\=\[\]\\\{\}\|\;\\'\:\"\,\.\/\<\>\?'), ErrorCode: 'notMatchingRegex', ErrorMessage: 'SpecialCharTest must match special character regex.' }
21157            ],
21158            'DefaultErrorMessageTest':[
21159                {Validator: RegexValidatorFactory('123'), ErrorCode: 'notMatchingRegex', ErrorMessage: 'Property {0} does not match required format.' }
21160            ]
21161        };
21162    }
```

The DSL script that causes this error is [Validations.rhe](https://github.com/Rhetos/Rhetos/blob/master/CommonConcepts/CommonConceptsTest/DslScripts/Validations.rhe), line 350: `RegExMatch` concept on `ShortString SpecialCharTest`.

It seems that the issue is caused by the combination of a backslash and a single quote. I am not sure, but the cause of the problem might be in handling of the backslash character when generating the typescript source code.

**CommonConceptsTest package** is part of the Rhetos framework repository. You can test it in AdminGui project by cloning or downloading the [Rhetos source](https://github.com/Rhetos/Rhetos) and adding the following line in your *RhetosPackages.config* file at AdminGui Rhetos server:

```xml
<package id="Rhetos.CommonConceptsTest" source="C:\... path to CommonConceptsTest folder in Rhetos framework repositoy" />
```

--- a/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Contants/RegularExpressions.cs
+++ b/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Contants/RegularExpressions.cs
@@ -10,5 +10,8 @@
         public const string DetectLastComma3 = "],\r\n        };\r\n";
         public const string DetectLastComma4 = "},\r\n        ];";
         public const string Decimal = "^-?[0-9]+(\\.[0-9]+)?$";
+        // Detect single quote if it is not escape yet and its replacement pattern
+        public const string SingleQuoteEscape = @"([^\\])'";
+        public const string SingleQuoteReplacement = @"$1\'";
     }
 }

--- a/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Generators/Validators/RegularExpressionValidatorGenerator.cs
+++ b/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Generators/Validators/RegularExpressionValidatorGenerator.cs
@@ -25,6 +25,7 @@ using Angular2ModelGenerator.Templates;
 using Rhetos.Dsl.DefaultConcepts;
 using Rhetos.Extensibility;
 using System.ComponentModel.Composition;
+using System.Text.RegularExpressions;
 
 namespace Angular2ModelGenerator.Generators.Validators
 {
@@ -35,7 +36,7 @@ namespace Angular2ModelGenerator.Generators.Validators
         protected override string GenerateCode(RegExMatchInfo info)
         {
             return ValidatorTemplates.Regex(
-                StringHelper.EscapeSpecialCharacters(info.RegularExpression, TypeScript.SpecialCharacters),
+                Regex.Replace(info.RegularExpression, RegularExpressions.SingleQuoteEscape, RegularExpressions.SingleQuoteReplacement),
                 StringHelper.EscapeSpecialCharacters(info.ErrorMessage, TypeScript.SpecialCharacters)
             );
         }

--- a/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Templates/FilterTemplates.cs
+++ b/Plugins/AdminGuiPlugin/Angular2ModelGenerator/Templates/FilterTemplates.cs
@@ -10,7 +10,7 @@
 
         public static string GetHasReadRowPermissions()
         {
-            return @"private getHasReadRowPermissions(DataStructure: IEmptyConstruct): boolean { return true; }
+            return @"public getHasReadRowPermissions(): boolean { return true; }
                 ";
         }
     }

